testctrl/svc/operations_server.go

```diff
@@ -16,29 +16,259 @@ package svc
+	emptypb "github.com/golang/protobuf/ptypes/empty"
+	timestamppb "github.com/golang/protobuf/ptypes/timestamp"
+	svcpb "github.com/grpc/grpc/testctrl/proto/scheduling/v1"
+	"github.com/grpc/grpc/testctrl/svc/store"
+	"github.com/grpc/grpc/testctrl/svc/types"
+	lrpb "google.golang.org/genproto/googleapis/longrunning"
+	codepb "google.golang.org/genproto/googleapis/rpc/code"
+	statuspb "google.golang.org/genproto/googleapis/rpc/status"
 )
 
-type OperationsServer struct{}
+// operationNamePrefix is the prefix present in all operation anmes.
+const operationNamePrefix = "operations/"
 
-func (o *OperationsServer) ListOperations(ctx context.Context, req *lrPb.ListOperationsRequest) (*lrPb.ListOperationsResponse, error) {
-	return nil, nil
+// OperationsServer implements the google.longrunning.Operations interface
+// to provide access to a data Store.
+type OperationsServer struct {
+	store store.Store
 }
 
-func (o *OperationsServer) GetOperation(ctx context.Context, req *lrPb.GetOperationRequest) (*lrPb.Operation, error) {
-	return nil, nil
+// NewOperationsServer constructs a new instance of OperationsServer.
+func NewOperationsServer(store store.Store) *OperationsServer {
+	return &OperationsServer{
+		store: store,
+	}
 }
 
-func (o *OperationsServer) DeleteOperation(ctx context.Context, req *lrPb.DeleteOperationRequest) (*emptyPb.Empty, error) {
-	return nil, nil
+// getSessionName converts an operation name to a session name.
+// This function returns an error if the operation name is invalid.
+func getSessionName(operationName string) (string, error) {
+	if !strings.HasPrefix(operationName, operationNamePrefix) {
+		return operationName, fmt.Errorf(
+			"invalid operation name: %q", operationName)
+	}
+	return strings.TrimPrefix(operationName, operationNamePrefix), nil
 }
 
-func (o *OperationsServer) CancelOperation(ctx context.Context, req *lrPb.CancelOperationRequest) (*emptyPb.Empty, error) {
-	return nil, nil
+// getOperationName converts a session name to an operation name.
+func getOperationName(sessionName string) string {
+	return fmt.Sprintf("%s%s", operationNamePrefix, sessionName)
 }
 
-func (o *OperationsServer) WaitOperation(ctx context.Context, req *lrPb.WaitOperationRequest) (*lrPb.Operation, error) {
-	return nil, nil
+// isOperationDone returns true if the operation is finished, and false
+// otherwise, based on the latest event received by the operation's session.
+func isOperationDone(e *types.Event) bool {
+	if e == nil {
+		return false
+	}
+	switch e.Kind {
+	case types.InternalErrorEvent:
+		return true
+	case types.DoneEvent:
+		return true
+	case types.ErrorEvent:
+		return true
+	default:
+		return false
+	}
+}
+
+// isOperationSuccess returns true if the operation is finished and is
+// successful, and false otherwise, based on the latest event received by
+// the operation's session.
+func isOperationSuccess(e *types.Event) bool {
+	return (isOperationDone(e) && e.Kind == types.DoneEvent)
```

Comment:
Do we need to check `isOperationDone` here?

Reply:
Yes. I may rethink this bug the idea is that isOperationDone(e) is false if e is nil. I may also rename this to isOperationFinished, so it is not confused with having a DoneEvent.